apps/frontend/src/lib/ws/index.ts

```diff
@@ -0,0 +1,104 @@
+import {BASE_API_WS} from "@/config";
+import {
+  WebSocketRequest,
+  WebSocketRequestTypes,
+  WebSocketResponseSchema,
+  WebSocketResponseTypes,
+  errorToString,
+  WebSocketResponse,
+  WebSocketErrorPayloadSchema,
+  WebSocketErrorPayload,
+  CreateMessageDataSchema,
+  CreateMessageData, MessageData, MapCreateMessageDataToMessageData
+} from "@types";
+import {getUserAndSession} from "@/lib/storage/localStorage/user";
+import {getUserSigPubKey} from "@/lib/user";
+import {storage} from "@/lib/storage";
+
+export const wsClient: WebSocket = new WebSocket(`${BASE_API_WS}`);
+
+// TODO: solve "TS2702: WebSocket only refers to a type, but is being used as a namespace here."
+wsClient.onopen = async (e: WebSocket.Event) => {
+  console.log("Open ws server connection")
+
+  if (e) {
+    console.log("Client onopen:", e)
+  }
+
+  //
+  const user = await storage.getUser();
+  const session = await storage.getSession();
+
+  if (user) {
+    const sigPubKey: string = getUserSigPubKey(user);
+    if (sigPubKey && session) {
+      // WS connection is established earlier but need user public key for any requests
+      wsConnectUser(session.authTokenValue, sigPubKey);
+    }
+  } else {
+    console.warn("No user public signing key available, unable to establish websocket connection")
+  }
+};
+
+// TODO: solve "TS2702: WebSocket only refers to a type, but is being used as a namespace here."
+wsClient.onmessage = async (e: WebSocket.MessageEvent) => {
+  if (!e || !e.data) {
+    console.warn("Message is invalid")
+    return;
+  }
+
+  try {
+    const resp: WebSocketResponse = WebSocketResponseSchema.parse(JSON.parse(e.data));
+
+    switch (resp.type) {
+      case WebSocketResponseTypes.MSG:
+        const createMessage: CreateMessageData = CreateMessageDataSchema.parse(resp.payload);
+
+        // Convert create to message by adding created at
+        const message: MessageData = MapCreateMessageDataToMessageData(createMessage)
+
+        if (message) {
+          await storage.processNewMessages([message]);
```

Comment:
I need to do a check in the refresh message to avoid double-saving the same messages.